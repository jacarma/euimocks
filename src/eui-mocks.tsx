import type * as euitypes from '@elastic/eui';
import { ReactNode } from 'react';
import { vi } from 'vitest';

// TODO: this is global, we could have problems with parallel tests
let mocksActivated = false;
export const activateEuiMocks = () => {
  mocksActivated = true;
};
export const deactivateEuiMocks = () => {
  mocksActivated = false;
};

export async function mockEui() {
  const eui = (await vi.importActual('@elastic/eui')) as typeof euitypes;

  type TypeOfEui = typeof eui;
  type KeysOfEui = keyof TypeOfEui;
  type ValuesOfEui = TypeOfEui[KeysOfEui];

  const mockedEui = { ...eui };
  Object.entries(mockedEui).forEach(([key, value]) => {
    const name = key as KeysOfEui;
    if (
      typeof name === 'string' &&
      name.startsWith('Eui') &&
      typeof value === 'function'
    ) {
      // TODO: the objective is to mock the UI components, not everything that starts with Eui
      mockedEui[name] = mockComponent(name, value) as never;
    }
  });

  function mockComponent(name: string, value: ValuesOfEui) {
    // this component returns the mocked component or the original one depending on the mocksActivated flag
    return function ({ children, ...props }: { children: ReactNode }) {
      if (mocksActivated) {
        return <MockedEuiComponent {...props}>{children}</MockedEuiComponent>;
      }
      return typeof value === 'function'
        ? value({ children, ...props }) // TODO: typescript warns us that we are treating everything as a component, they are right
        : value;
    };

    // this is the component we render instead when mocks are active
    function MockedEuiComponent({
      children,
      ...props
    }: {
      children: ReactNode;
    }) {
      return (
        <div role={name} {...props}>
          {children}
        </div>
      );
    }
  }
  return { ...mockedEui };
}
