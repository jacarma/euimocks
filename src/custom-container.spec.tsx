import { vi } from 'vitest'; // I'm using vitest, but jest should work very similar
import { mockEui, activateEuiMocks, deactivateEuiMocks } from './eui-mocks';
vi.doMock('@elastic/eui', mockEui); // this needs to be done before importing the components
import CustomContainer from './custom-container';
import renderer from 'react-test-renderer';
import { EuiButton } from '@elastic/eui';

const nestedCustomAndEuiComponents = (
  <CustomContainer>
    <CustomContainer>Hello World</CustomContainer>
  </CustomContainer>
);

const button = (
  <EuiButton color="primary" fill>
    I'm a button
  </EuiButton>
);

describe('EuiButton', async () => {
  it('serialize with mocks', () => {
    // this snapshot shows <div role="EuiButton" color="primary" fill>I'm a button</div>
    // just the component name, props and children)
    activateEuiMocks();
    expect(renderer.create(button).toJSON()).toMatchSnapshot();
  });
  it('serialize without mocks', () => {
    // this snapshows shows <button className="..." disabled={false} type="button"><span className="..."><span className="eui-textTruncate">I'm a button</span></span></button>
    // the whole rendered component
    deactivateEuiMocks();
    expect(renderer.create(button).toJSON()).toMatchSnapshot();
  });
});

describe('custom-container', async () => {
  // we need to include the children in the mocks or we loose anything iside the EuiPanel
  // everything rendered inside the custom component is rendered completely. But the Eui components are mocked
  it('serialize with mocks', () => {
    activateEuiMocks();
    expect(
      renderer.create(nestedCustomAndEuiComponents).toJSON()
    ).toMatchSnapshot();
  });
  it('serialize without mocks', () => {
    deactivateEuiMocks();
    expect(
      renderer.create(nestedCustomAndEuiComponents).toJSON()
    ).toMatchSnapshot();
  });
});
