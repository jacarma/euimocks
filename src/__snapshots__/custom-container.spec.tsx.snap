// Vitest Snapshot v1

exports[`EuiButton > serialize with mocks 1`] = `
<div
  color="primary"
  fill={true}
  role="EuiButton"
>
  I'm a button
</div>
`;

exports[`EuiButton > serialize without mocks 1`] = `
<button
  className="euiButton css-1u1unpk-euiButtonDisplay-m-defaultMinWidth-m-fill-primary"
  disabled={false}
  type="button"
>
  <span
    className="css-1km4ln8-euiButtonDisplayContent"
  >
    <span
      className="eui-textTruncate"
    >
      I'm a button
    </span>
  </span>
</button>
`;

exports[`custom-container > serialize with mocks 1`] = `
<div
  role="EuiPanel"
>
  <div
    role="EuiText"
  >
    <h1>
      Custom Component
    </h1>
    <p>
      Some text
    </p>
  </div>
  <div
    color="subdued"
    role="EuiPanel"
  >
    <div
      role="EuiPanel"
    >
      <div
        role="EuiText"
      >
        <h1>
          Custom Component
        </h1>
        <p>
          Some text
        </p>
      </div>
      <div
        color="subdued"
        role="EuiPanel"
      >
        Hello World
      </div>
      <div
        color="primary"
        fill={true}
        role="EuiButton"
      >
        Button at the end
      </div>
    </div>
  </div>
  <div
    color="primary"
    fill={true}
    role="EuiButton"
  >
    Button at the end
  </div>
</div>
`;

exports[`custom-container > serialize without mocks 1`] = `
<div
  className="euiPanel euiPanel--plain euiPanel--paddingMedium css-1yz5lbp-euiPanel-grow-m-m-plain-hasShadow"
>
  <div
    className="euiText css-f0maoj-euiText-m"
  >
    <h1>
      Custom Component
    </h1>
    <p>
      Some text
    </p>
  </div>
  <div
    className="euiPanel euiPanel--subdued euiPanel--paddingMedium css-isico0-euiPanel-grow-m-m-subdued"
  >
    <div
      className="euiPanel euiPanel--plain euiPanel--paddingMedium css-1yz5lbp-euiPanel-grow-m-m-plain-hasShadow"
    >
      <div
        className="euiText css-f0maoj-euiText-m"
      >
        <h1>
          Custom Component
        </h1>
        <p>
          Some text
        </p>
      </div>
      <div
        className="euiPanel euiPanel--subdued euiPanel--paddingMedium css-isico0-euiPanel-grow-m-m-subdued"
      >
        Hello World
      </div>
      <button
        className="euiButton css-1u1unpk-euiButtonDisplay-m-defaultMinWidth-m-fill-primary"
        disabled={false}
        type="button"
      >
        <span
          className="css-1km4ln8-euiButtonDisplayContent"
        >
          <span
            className="eui-textTruncate"
          >
            Button at the end
          </span>
        </span>
      </button>
    </div>
  </div>
  <button
    className="euiButton css-1u1unpk-euiButtonDisplay-m-defaultMinWidth-m-fill-primary"
    disabled={false}
    type="button"
  >
    <span
      className="css-1km4ln8-euiButtonDisplayContent"
    >
      <span
        className="eui-textTruncate"
      >
        Button at the end
      </span>
    </span>
  </button>
</div>
`;
